### ISSUE TITLE
Ticker stops functioning when browser tab isn't active

### ISSUE THREAD

**Issue by AllanPooley** [AndreasFaust/react-ticker#10]
Hey @AndreasFaust thanks for creating this package.

I'm using it on a 404 page for a site I built: https://www.wagec.org.au/404

It works great:

![image](https://user-images.githubusercontent.com/9204586/69018552-1c0bc580-0a01-11ea-9ea3-94664d7e57f4.png)

The only slightly odd behaviour is that the animation / rendering stops when the tab isn't visible on the screen, creating situations like this:

<img width="2048" alt="Screen Shot 2019-11-18 at 12 42 46 pm" src="https://user-images.githubusercontent.com/9204586/69018598-3d6cb180-0a01-11ea-90c9-8fdf5b0e34c3.png">

I'm guessing this is for performance reasons? Is there any way to prevent this?

**Comment by ololoepepe** [comment]
**TL;DR** _(It seems to me that)_ there is no quick and easy way to fix this.
It must be something to do with `setTimeout`-like functions. As I can see, here https://github.com/AndreasFaust/react-ticker/blob/master/src/Element.js#L168 `requestAnimationFrame` is used, which is stopped when a window or tab blurs, see https://stackoverflow.com/questions/15871942/how-do-browsers-pause-change-javascript-when-tab-or-window-is-not-active
When the tab is focused again, the `step` callback passed to RAF is finally triggered. But! Look here: https://github.com/AndreasFaust/react-ticker/blob/master/src/Element.js#L178 A timestamp is used to calculate the offset. This timestamp is something like `Date.now() - previousTimestamp` multiplied by some coefficient, which gives us the offset. Obviously, the difference is large enough when we put a tab to background for a long time, so the offset is large and we get a gap.
The code looks a bit too complicated to me, you may try to work around this bug, but I believe it's better to re-implement. As I'm also interested in such a package, I'll try to find a replacement or implement it myself in not-so-far future. I'll try not to forget to let you know then.

**Comment by AndreasFaust** [comment]
Have you tried to completely destroy and re-initialize `react-ticker` when the user comes back to the tab?
Use this API: https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API

**Comment by ololoepepe** [comment]
@AndreasFaust Should we? Why not just use `.animate` for animation and `setTimeout` for checking/calling `triggerNext`?

**Comment by AndreasFaust** [comment]
If you mean this API: https://developer.mozilla.org/en-US/docs/Web/API/Element/animate
It is not supported by all browsers yet: https://caniuse.com/#search=element.animate

**Comment by ololoepepe** [comment]
@AndreasFaust Polyfills exist for that purpose. Or, if it seems an overkill, just insert animation with some generated unique id into a style tag.

Well, OK, at least we can encapsulate that visibility API stuff in the lib. That's a 100% common case, so it should be handled.

**Comment by AndreasFaust** [comment]
Thank you for the suggestions. Unfortunately I have other priorities at the moment. If you’ll find the time, I am looking forward to receive a pull request from you, which will be merged quickly.

**Comment by AllanPooley** [comment]
Thanks for the input @ololoepepe @AndreasFaust

**Comment by mrclay** [comment]
Not rendering the ticker when the document is made hidden indeed works great.

I’m hoping to release a useDocVisible hook for this purpose soon.

**Comment by thriceAmbo** [comment]
Got around this issue by using https://www.npmjs.com/package/react-page-visibility

I'm using React Hooks but easy to update if you're not using them.

Updating original example from docs:

```
import React, { useState } from 'react'
import Ticker from 'react-ticker'
import PageVisibility from 'react-page-visibility'

const MoveStuffAround = () => {
  const [pageIsVisible, setPageIsVisible] = useState(true)

  const handleVisibilityChange = (isVisible) => {
    setPageIsVisible(isVisible)
  }

  return (
    <PageVisibility onChange={handleVisibilityChange}>
      {pageIsVisible && (
        <Ticker>
          {({ index }) => (
              <>
                  <h1>This is the Headline of element #{index}!</h1>
                  <img src="www.my-image-source.com/" alt=""/>
              </>
          )}
        </Ticker>
      )}
    </PageVisibility>
  )
}

export default MoveStuffAround
```

**Comment by AllanPooley** [comment]
Thanks @thriceAmbo, this work around works!

**Comment by AndreasFaust** [comment]
Thanks all of you for working this out! 
I will remark this workaround in the docs and consider it for my planned rewrite of `react-ticker`.

**Comment by ololoepepe** [comment]
@thriceAmbo One should use `blur`/`focus` events instead of `visibilitychange`. `visibilitychange` only fires when switching between tabs, but does not fire when the tab is active and you put your browser to background. `blur`/`focus` do fire in both scenarios.
